File: Algorithms/Python/Symmetric-Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):

    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def comparelr(nodeA, nodeB):
            if not nodeA and not nodeB:
                return True
            elif nodeA and nodeB:
                return (nodeA.val == nodeB.val) and comparelr(nodeA.left, nodeB.right) and comparelr(nodeA.right, nodeB.left)
            else:
                return False
        if not root:
            return True
        return comparelr(root.left, root.right)

File: Algorithms/Java/Symmetric Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isSymmetric(TreeNode root) {
		if (root == null)
			return true;
		return isMirror(root.left, root.right);
	}

	public boolean isMirror(TreeNode treeA, TreeNode treeB) {
		if (treeA == null && treeB == null)
			return true;
		if (treeA == null || treeB == null)
			return false;
		if (treeA.val == treeB.val)
			return isMirror(treeA.left, treeB.right)
					&& isMirror(treeA.right, treeB.left);
		return false;
	}
}